### PR TITLE
Added Locks wherever LiteDB is used.

### PR DESCRIPTION
**Pull request by manas-droid**
LiteDB fails to handle Multiple threads accessing .db files. Therefore, had to manually add locks.

Thank you for your contribution.
Before submitting this PR, please make sure:

- [ ] PR description and commit message should describe the changes done in this PR
- [ ] Verify the PR is pointing to correct branch i.e. Release or Beta branch if the code fix is for specific release , else point it to master
- [ ] Latest Code from master or specific release branch is merged to your branch
- [ ] No unwanted\commented\junk code is included
- [ ] No new warning upon build solution
- [ ] Code Summary\Comments are added to my code which explains what my code is doing
- [ ] Existing unit test cases are passed
- [ ] New Unit tests are added for your development
- [ ] Sanity Tests are successfully executed for New and Existing Functionality
- [ ] Verify that changes are compatible with all relevant browsers and platforms.
- [ ] After creating pull request there should not be any conflicts
- [ ] Resolve all Codacy comments
- [ ] Builds and checks are passed before PR is sent for review
- [ ] Resolve code review comments
- [ ] Update the Help Library document to match any feature changes

## Summary by CodeRabbit

- **New Features**
  - Enhanced database access to ensure thread safety and prevent data conflicts.
  
- **Bug Fixes**
  - Implemented improved error handling and logging during database operations.

- **Refactor**
  - Optimized database connection management for better performance and reliability.

### CODE REVIEW

File: Ginger/GingerCoreNET/LiteDBFolder/LiteDbManager.cs [hunk 1]
```diff
@@ -138,11 +139,15 @@ public void WriteToLiteDb()
...
         }
+        private static readonly SemaphoreSlim semaphore = new(1,1);
         public void WriteToLiteDb(string reportLevelName, List<LiteDbReportBase> objectColl)
         {
+            semaphore.Wait();
             var objectLiteColl = GetObjectLiteData(reportLevelName);
             //var objectColl = this.GetGingerObject(reportLevelName, gingerObject);
             dbConnector.SetCollection(objectLiteColl, objectColl);
+            semaphore.Release();
+
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbManager.cs]: You are right my lord

File: Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs [hunk 2]
```diff
@@ -128,10 +132,7 @@ public void SetCollection<T>(ILiteCollection<T> baseColl, List<T> updateData)
         {
             try
             {
-                using (var db = new LiteDatabase(this.ConnectionString))
-                {
-                    baseColl.Upsert(updateData);
-                }
+               baseColl.Upsert(updateData);
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs:135]: using was removed as the object of LiteDatabase class was never used in the function.

File: Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs [hunk 3]
```diff
@@ -39,19 +41,21 @@ public EntityBuilder<T> GetMapper<T>()
...
-            {
-
-            }
-            return collection;
+                catch (Exception ex)
+                {
+                    Reporter.ToLog(eLogLevel.ERROR, $"Failed to Get Collection: {collectionName}", ex);
+                }
+             
+                return collection;
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs:58]: @Maheshkale447 should I rethrow the exception here?

File: Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs [hunk 4]
```diff
@@ -22,11 +22,13 @@
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Threading;
 
 namespace Amdocs.Ginger.CoreNET.LiteDBFolder
 {
     public class LiteDbConnector
     {
+        private static readonly object readLock = new();
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs:31]: Ok , will remove this.

File: Ginger/GingerCoreNET/DataSource/LiteDB.cs [hunk 5]
```diff
@@ -62,34 +65,47 @@ public override void AddColumn(string tableName, string columnName, string colum
             {
                 Reporter.ToUser(eUserMsgKey.StaticErrorMessage, "Please enter valid column name");
             }
+            finally
+            {
+                LiteDBLock.Wait();
+            }
```
[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/DataSource/LiteDB.cs]: Thank you!

File: Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Maheshkale447) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs:60]: Yes, either rethrow it or remove the try-catch block

[the remark this answers is not in the document]
Comment (manas-droid) [reply Ginger/GingerCoreNET/LiteDBFolder/LiteDbConnector.cs:60]: ok fine, Thanks!

**Review by Maheshkale447 — APPROVED** [review]

**Pull request merged**